cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/LocalStreamService.java

```diff
@@ -70,41 +70,76 @@ protected void initialize() throws Exception {
     for (StreamSpecification streamSpec : streamMetaStore.listStreams()) {
       StreamConfig config = streamAdmin.getConfig(streamSpec.getName());
       long filesSize = StreamUtils.fetchStreamFilesSize(config);
-      aggregators.put(streamSpec.getName(), new Aggregator(streamSpec.getName(), filesSize));
+      createSizeAggregator(streamSpec.getName(), filesSize);
     }
   }
 
   @Override
   protected void doShutdown() throws Exception {
-    // No-op
+    for (StreamSizeAggregator streamSizeAggregator : aggregators.values()) {
+      streamSizeAggregator.cancel();
+    }
   }
 
   @Override
   protected void runOneIteration() throws Exception {
     // Get stream size - which will be the entire size - and send a notification if the size is big enough
     for (StreamSpecification streamSpec : streamMetaStore.listStreams()) {
-      Aggregator aggregator = aggregators.get(streamSpec.getName());
-      if (aggregator == null) {
+      StreamSizeAggregator streamSizeAggregator = aggregators.get(streamSpec.getName());
+      if (streamSizeAggregator == null) {
         // First time that we see this Stream here
-        aggregator = new Aggregator(streamSpec.getName(), 0);
-        aggregators.put(streamSpec.getName(), aggregator);
+        streamSizeAggregator = createSizeAggregator(streamSpec.getName(), 0);
       }
-      aggregator.checkAggregatedSize();
+      streamSizeAggregator.checkAggregatedSize();
     }
     isInit = false;
   }
 
   /**
+   * Create a new aggregator for the {@code streamName}, and add it to the existing map of {@link Cancellable}
+   * {@code aggregators}. This method does not cancel previously existing aggregator associated to the
+   * {@code streamName}.
+   *
+   * @param streamName stream name to create a new aggregator for
+   * @param baseCount stream size from which to start aggregating
+   * @return the created {@link StreamSizeAggregator}
+   */
+  private StreamSizeAggregator createSizeAggregator(String streamName, long baseCount) {
+
+    // Handle stream truncation, by creating creating a new empty aggregator for the stream
+    // and cancelling the existing one
+    final Cancellable truncationSubscription =
+      getStreamCoordinatorClient().addListener(streamName, new StreamPropertyListener() {
+        @Override
+        public void generationChanged(String streamName, int generation) {
+          Cancellable previousAggregator = aggregators.replace(streamName, createSizeAggregator(streamName, 0));
+          if (previousAggregator != null) {
+            previousAggregator.cancel();
+          }
+        }
+      });
+
+    StreamSizeAggregator newAggregator = new StreamSizeAggregator(streamName, baseCount) {
+      @Override
+      public void cancel() {
```

Comment:
Just pass the Cancellable to aggregator, no need to do abstract class and override.